design/implicit-boot-mode.md

```diff
@@ -0,0 +1,197 @@
+right mode at the right time.
+
+The first approach would be to try to always require the same boot
+mode for all hosts. Most modern x86 hosts support UEFI. It is
+therefore useful as a default. However, many contributors have
+somewhat older hardware, which works better with the BIOS boot
+mode. While this hardware is not necessarily our target, it isn't
+practical to abandon support entirely by always requiring
+UEFI. Similarly, there are reasons we cannot always require a host to
+use the BIOS boot mode. UEFI supports desirable features, such as
+secure booting, non-x86 hardware, and PXE on IPv6 networks. For these
+reasons, a single boot mode will not work.
+
+The desired features do give us some hints about how we can determine
+the boot mode implicitly, however. For example, we have no plans to
+support virtual media with IPMI. We can also tell whether the
+provisioning network is using IPv4 or IPv6. Using this information, we
+could build a set of rules to determine the boot mode. However, we
+would have to document those rules, and they may not seem obvious to
+the end-user.
+
+At least some of the features are tied to the protocol used to
+communicate with the BMC. We could therefore link the boot mode to the
+BMC driver, and have a deterministic method for choosing the boot mode
+that is also easy to document.
+
+The proposal, therefore, is to always use BIOS mode when using IPMI
+and to prefer UEFI when using Redfish. We can implement this by adding
+a new method to the [AccessDetails
+interface](https://github.com/metal3-io/baremetal-operator/blob/master/pkg/bmc/access.go#L27)
+to return the properties for a host, including the `boot_mode`. This
+API is consistent with other similar methods in that class that return
+other sets of data passed to Ironic when a host is registered.
+
+We can be somewhat flexible with the rules by allowing the BMC driver
+in metal3 to look at some of the other settings for the host to decide
+which boot mode to set. For example, when the driver author knows that
+Ironic supports determining the boot mode correctly, it could choose
+to skip setting a mode unless the other settings (use of virtual media
+or IPv6, for example) require a particular value. This would
+complicate the documentation, but may be necessary for some drivers
+where trying to have Ironic set the boot mode frequently causes
+provisioning failures and it is better to use the host's existing
+configuration when possible.
+
+### User Stories [optional]
+
+#### As a user, I want to provision a host without using virtual media instead of PXE
+
+#### As a user, I want to provision a host that uses an IPMI-based BMC
+
+#### As a user, I want to provision a host using an IPv6 provisioning network
+
+### Risks and Mitigations
+
+## Design Details
+
+- Update the [AccessDetails
+  interface](https://github.com/metal3-io/baremetal-operator/blob/master/pkg/bmc/access.go#L27)
+  to include the new method, `NodeProperties()`.
```

Comment:
This method is supposed to return a structure? Like you mentioned earlier, node ```properties```. Where would be the best place to put such a structure? The same ```AccessDetails``` interface? (Would make sense, since this is where we generally handle vendor-specific stuff.

Reply:
See the earlier response about using a map.

Reply:
Got it. Thanks.